testing-guidelines/ci.md

```diff
@@ -0,0 +1,431 @@
+
+### Context
+
+This is some additional context about why things are the way they are and the challenges we had to overcome. This isn't essential reading  but may be helpful if you want to understand why we chose the current design. This will also be helpful as we iterate further on CI in the future; the trauma is fresh as I write this but it will fade with time, and so too will my memory of it.
+
+If you haven't read the previous sections, this part may not make complete sense.
+
+#### The problem
+
+We have the following constraints and challenges:
+
+- We have a relatively comprehensive CI suite, different components have their own idiosyncracies.
+- Many of our jobs need access to secrets but security is a high priority.
+- We are an open source project and want the same experience for contributors (PRs from forks) as the core team have (PRs from repo branches).
+- We want to make all of the important checks required.
+- We want CI to be fast, certain jobs should be skipped where appropriate. These jobs may still be 'required' to pass if—and only if—they run.
+- CI should give confidence over time, we don't want to be overlooking the odd ❌ as a 'normal' part of our CI, even for optional jobs. This will erode trust.
+- Many of our CI jobs share the same steps. We want to avoid excessive duplication where possible for maintenance reasons.
+
+Some of these are discrete problems with their own discrete solutions but a lot of the challenges stem from when certain GitHub Action events occur and how priveleged/ secure those event 'types' are.
+
+#### Demystifying event triggers
+
+Workflows are a discreet set of jobs with a discrete set of steps. It might be reasonable to assume that a workflow is a workflow. Sadly this isn't true, the event that triggers the workflow dictates not only when that workflow will run (which makes sense) but also a bunch of other information about both its environment and even which version of that workflow file will run (this is a git repo after all). This latter feature, _also_ makes sense, but it isn't immediately apparent.
+
+- `pull_request` - This event runs correctly on contributor PRs and check out the correct branch by default (more on this later) but it doesn't not have access to secrets.
+- `pull_request_target` - Same as `pull_request` but it _does_ have access to secrets. However because this event runs in to context of the branch (and repo) the PR is made from, that PR has also has direct access to secrets making it insecure.
+- `push` - When triggered from a fork, this will essentially be a `push` to the fork not the target repo. In many cases the workflow won't run ata ll (depends on the settings of the fork) and it won't have access to secrets from the base repo. Even if it did, it would be inscure.
```

Comment:
```suggestion
- `push` - When triggered from a fork, this will essentially be a `push` to the fork, not the target repo. In many cases the workflow won't run at all (depends on the settings of the fork) and it won't have access to secrets from the base repo. Even if it did, it would be insecure.
```